File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/ScryptParams.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bccrypto.scrypt
import com.blockchaincommons.dcbor.Cbor

/**
 * Scrypt-based key derivation parameters.
 *
 * CDDL:
 * ```
 * ScryptParams = [2, Salt, log_n: uint, r: uint, p: uint]
 * ```
 *
 * @property salt the salt used for key derivation
 * @property logN the log2 of the CPU/memory cost parameter
 * @property r the block size parameter
 * @property p the parallelisation parameter
 */
class ScryptParams(
    val salt: Salt,
    val logN: Int = DEFAULT_LOG_N,
    val r: Int = DEFAULT_R,
    val p: Int = DEFAULT_P,
) : KeyDerivation {

    /** Creates default parameters with a random 16-byte salt and standard cost factors. */
    constructor() : this(Salt.createWithLength(SALT_LEN), DEFAULT_LOG_N, DEFAULT_R, DEFAULT_P)

    override fun lock(contentKey: SymmetricKey, secret: ByteArray): EncryptedMessage {
        val derivedKey = deriveKey(secret)
        val encodedMethod = toCbor().toCborData()
        return derivedKey.encrypt(contentKey.data(), encodedMethod, null)
    }

    override fun unlock(encryptedMessage: EncryptedMessage, secret: ByteArray): SymmetricKey {
        val derivedKey = deriveKey(secret)
        val decrypted = derivedKey.decrypt(encryptedMessage)
        return SymmetricKey.fromData(decrypted)
    }

    private fun deriveKey(secret: ByteArray): SymmetricKey {
        val derived = scrypt(secret, salt.asBytes(), 32, logN, r, p)
        return SymmetricKey.fromData(derived)
    }

    /** Encodes these parameters as a CBOR array. */
    fun toCbor(): Cbor = Cbor.fromArray(listOf(
        Cbor.fromInt(KeyDerivationMethod.Scrypt.index),
        salt.untaggedCbor(),
        Cbor.fromInt(logN),
        Cbor.fromInt(r),
        Cbor.fromInt(p),
    ))

    override fun toString(): String = "Scrypt"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ScryptParams) return false
        return salt == other.salt &&
            logN == other.logN &&
            r == other.r &&
            p == other.p
    }

    override fun hashCode(): Int {
        var result = salt.hashCode()
        result = 31 * result + logN
        result = 31 * result + r
        result = 31 * result + p
        return result
    }

    companion object {
        const val SALT_LEN = 16
        const val DEFAULT_LOG_N = 15
        const val DEFAULT_R = 8
        const val DEFAULT_P = 1

        /**
         * Decodes [ScryptParams] from a CBOR array.
         *
         * The expected format is `[2, salt_bytes, log_n, r, p]`.
         */
        fun fromCbor(cbor: Cbor): ScryptParams {
            val a = cbor.tryArray()
            if (a.size != 5) {
                throw BcComponentsException.general("Invalid ScryptParams: expected 5 elements, got ${a.size}")
            }
            // a[0] is the method index (already consumed by caller)
            val salt = Salt.fromUntaggedCbor(a[1])
            val logN = a[2].tryInt()
            val r = a[3].tryInt()
            val p = a[4].tryInt()
            return ScryptParams(salt, logN, r, p)
        }
    }
}

File: csharp/BCComponents/BCComponents/ScryptParams.cs
using BlockchainCommons.BCCrypto;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// Scrypt-based key derivation parameters.
/// </summary>
/// <remarks>
/// CDDL:
/// <code>
/// ScryptParams = [2, Salt, log_n: uint, r: uint, p: uint]
/// </code>
/// </remarks>
public sealed class ScryptParams : IEquatable<ScryptParams>, IKeyDerivation
{
    /// <summary>Default salt length in bytes.</summary>
    public const int SaltLen = 16;

    /// <summary>Default log2 of the CPU/memory cost parameter.</summary>
    public const int DefaultLogN = 15;

    /// <summary>Default block size parameter.</summary>
    public const int DefaultR = 8;

    /// <summary>Default parallelisation parameter.</summary>
    public const int DefaultP = 1;

    /// <summary>Gets the salt used for key derivation.</summary>
    public Salt Salt { get; }

    /// <summary>Gets the log2 of the CPU/memory cost parameter.</summary>
    public int LogN { get; }

    /// <summary>Gets the block size parameter.</summary>
    public int R { get; }

    /// <summary>Gets the parallelisation parameter.</summary>
    public int P { get; }

    /// <summary>
    /// Creates Scrypt parameters with the given values.
    /// </summary>
    /// <param name="salt">The salt for key derivation.</param>
    /// <param name="logN">The log2 of the CPU/memory cost parameter.</param>
    /// <param name="r">The block size parameter.</param>
    /// <param name="p">The parallelisation parameter.</param>
    public ScryptParams(Salt salt, int logN = DefaultLogN, int r = DefaultR, int p = DefaultP)
    {
        Salt = salt;
        LogN = logN;
        R = r;
        P = p;
    }

    /// <summary>
    /// Creates default Scrypt parameters with a random 16-byte salt and standard
    /// cost factors.
    /// </summary>
    public ScryptParams()
        : this(Salt.CreateWithLength(SaltLen), DefaultLogN, DefaultR, DefaultP)
    {
    }

    /// <inheritdoc/>
    public EncryptedMessage Lock(SymmetricKey contentKey, byte[] secret)
    {
        var derivedKey = DeriveKey(secret);
        var encodedMethod = ToCbor().ToCborData();
        return derivedKey.Encrypt(contentKey.Data, encodedMethod, null);
    }

    /// <inheritdoc/>
    public SymmetricKey Unlock(EncryptedMessage encryptedMessage, byte[] secret)
    {
        var derivedKey = DeriveKey(secret);
        var decrypted = derivedKey.Decrypt(encryptedMessage);
        return SymmetricKey.FromData(decrypted);
    }

    private SymmetricKey DeriveKey(byte[] secret)
    {
        var derived = ScryptKdf.DeriveOpt(secret, Salt.AsBytes(), 32, (byte)LogN, (uint)R, (uint)P);
        return SymmetricKey.FromData(derived);
    }

    /// <summary>Encodes these parameters as a CBOR array.</summary>
    public Cbor ToCbor() => Cbor.FromList(new List<Cbor>
    {
        Cbor.FromInt((int)KeyDerivationMethod.Scrypt),
        Salt.UntaggedCbor(),
        Cbor.FromInt(LogN),
        Cbor.FromInt(R),
        Cbor.FromInt(P),
    });

    /// <summary>Decodes <see cref="ScryptParams"/> from a CBOR array.</summary>
    /// <param name="cbor">The CBOR array value.</param>
    /// <returns>A new <see cref="ScryptParams"/>.</returns>
    public static ScryptParams FromCbor(Cbor cbor)
    {
        var a = cbor.TryIntoArray();
        if (a.Count != 5)
            throw BCComponentsException.General($"Invalid ScryptParams: expected 5 elements, got {a.Count}");
        var salt = Salt.FromUntaggedCbor(a[1]);
        var logN = (int)a[2].TryIntoUInt64();
        var r = (int)a[3].TryIntoUInt64();
        var p = (int)a[4].TryIntoUInt64();
        return new ScryptParams(salt, logN, r, p);
    }

    /// <inheritdoc/>
    public override string ToString() => "Scrypt";

    /// <inheritdoc/>
    public bool Equals(ScryptParams? other)
    {
        if (other is null) return false;
        return Salt == other.Salt
            && LogN == other.LogN
            && R == other.R
            && P == other.P;
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is ScryptParams p && Equals(p);

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(Salt, LogN, R, P);
}
